service/keychain/cri/v1alpha/cri.go

```diff
@@ -0,0 +1,163 @@
+/*
+   Copyright The Soci Snapshotter Authors.
+
+   Licensed under the Apache License, Version 2.0 (the "License");
+   you may not use this file except in compliance with the License.
+   You may obtain a copy of the License at
+
+       http://www.apache.org/licenses/LICENSE-2.0
+
+   Unless required by applicable law or agreed to in writing, software
+   distributed under the License is distributed on an "AS IS" BASIS,
+   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+   See the License for the specific language governing permissions and
+   limitations under the License.
+*/
+
+/*
+   Copyright The containerd Authors.
+
+   Licensed under the Apache License, Version 2.0 (the "License");
+   you may not use this file except in compliance with the License.
+   You may obtain a copy of the License at
+
+       http://www.apache.org/licenses/LICENSE-2.0
+
+   Unless required by applicable law or agreed to in writing, software
+   distributed under the License is distributed on an "AS IS" BASIS,
+   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+   See the License for the specific language governing permissions and
+   limitations under the License.
+*/
+
+package crialpha
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"sync"
+	"time"
+
+	"github.com/awslabs/soci-snapshotter/service/resolver"
+	"github.com/containerd/containerd/log"
+	"github.com/containerd/containerd/reference"
+	distribution "github.com/containerd/containerd/reference/docker"
+	runtime_alpha "github.com/containerd/containerd/third_party/k8s.io/cri-api/pkg/apis/runtime/v1alpha2"
+)
+
+// NewCRIAlphaKeychain provides creds passed through CRI PullImage API.
+// This also returns a CRI image service server that works as a proxy backed by the specified CRI service.
+// This server reads all PullImageRequest and uses PullImageRequest.AuthConfig for authenticating snapshots.
+func NewCRIAlphaKeychain(ctx context.Context, connectCRI func() (runtime_alpha.ImageServiceClient, error)) (resolver.Credential, runtime_alpha.ImageServiceServer) {
+	server := &instrumentedAlphaService{config: make(map[string]*runtime_alpha.AuthConfig)}
+	go func() {
+		log.G(ctx).Debugf("Waiting for CRI service is started...")
+		for i := 0; i < 100; i++ {
+			client, err := connectCRI()
+			if err == nil {
+				server.criMu.Lock()
+				server.cri = client
+				server.criMu.Unlock()
+				log.G(ctx).Info("connected to backend CRI service")
+				return
+			}
+			log.G(ctx).WithError(err).Warnf("failed to connect to CRI")
+			time.Sleep(10 * time.Second)
+		}
+		log.G(ctx).Warnf("no connection is available to CRI")
```

Comment:
nit (kinda): wouldn't failing to connect to the CRI warrant more than just a warning?

Reply:
Ah, good catch. I've moved the `getCriConn` [method](https://github.com/awslabs/soci-snapshotter/blob/fb71740fa2e428127defdb7a032589ae33e2f195/cmd/soci-snapshotter-grpc/main.go#L146) outside of these callbacks, which isn't supposed to be the case, hence the log message here. I'll move it to be inside of each callback. 

In terms of whether or not we should error out here (eg: `Fatalf`), I don't think we do. There are multiple methods of getting credentials through to SOCI in K8s. If we can't establish a connection to CRI after set of retries, we should continue in hopes that we can get creds to SOCI through some other method later, either through the docker authstore or kubeconfig if its configured. That being said, I do see the case for having it emit an `Errorf` instead.